File: secretmanager/v1/python/google-cloud-secret-manager/secret-manager-client-secret-create-notifications.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# [START secretmanager_v1_secretmanagerservice_secret_create_notifications_with_topics]
import google.api_core.exceptions
from google.cloud import secretmanager_v1


def create_secret_with_notifications(
    project_id: str, secret_id: str, topic_id: str
) -> None:
    """Creates a new secret resource configured with Pub/Sub notifications. This demonstrates how to receive notifications when secret versions are added or destroyed.

    Args:
        project_id: Google Cloud Project ID (such as 'example-project-id')
        secret_id: ID of the secret to create (such as 'my-secret-id')
        topic_id: ID of the Pub/Sub topic (such as 'my-topic-id')
    """
    client = secretmanager_v1.SecretManagerServiceClient()

    parent = f"projects/{project_id}"

    try:
        response = client.create_secret(
            request={
                "parent": parent,
                "secret_id": secret_id,
                "secret": {
                    "replication": {"automatic": {}},
                    "topics": [{"name": f"projects/{project_id}/topics/{topic_id}"}],
                },
            }
        )

        print(f"Created secret: {response.name}")
        for topic in response.topics:
            print(f"  Topic: {topic.name}")

    except google.api_core.exceptions.AlreadyExists:
        print(
            f"Error: The secret '{secret_id}' already exists in project "
            f"'{project_id}'. Use a different, unique secret ID."
        )
    except Exception as e:
        print(f"An unexpected error occurred: {e}")


# [END secretmanager_v1_secretmanagerservice_secret_create_notifications_with_topics]

File: secretmanager/v1/nodejs/google-cloud-secret-manager/secret-manager-client-secret-create-notifications.js
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// [START secretmanager_v1_secretmanagerservice_secret_create_notifications_with_topics]

'use strict';

const {SecretManagerServiceClient} = require('@google-cloud/secret-manager');
const {status} = require('@grpc/grpc-js');

const client = new SecretManagerServiceClient();

/**
 * Create Secret with Pub/Sub Notifications.
 *
 * Creates a new secret resource configured with Pub/Sub notifications. This
 * sample demonstrates how to publish notifications when secret versions are
 * added or destroyed.
 *
 * @param projectId Google Cloud Project ID (such as 'example-project-id')
 * @param secretId ID of the secret to create (such as 'my-secret-id')
 * @param topicId ID of the Pub/Sub topic (such as 'my-topic-id')
 */
async function createSecretWithNotifications(projectId, secretId, topicId) {
  const parent = `projects/${projectId}`;

  try {
    const [secret] = await client.createSecret({
      parent,
      secretId,
      secret: {
        replication: {
          automatic: {},
        },
        topics: [
          {
            name: `projects/${projectId}/topics/${topicId}`,
          },
        ],
      },
    });

    console.log(`Created secret: ${secret.name}`);
    for (const topic of secret.topics) {
      console.log(`  Topic: ${topic.name}`);
    }
  } catch (err) {
    if (err.code === status.ALREADY_EXISTS) {
      console.error(
        `Error: The secret '${secretId}' already exists. Use a different secret ID.`,
      );
    } else {
      console.error('An unexpected error occurred:', err);
    }
  }
}

// [END secretmanager_v1_secretmanagerservice_secret_create_notifications_with_topics]

module.exports = {createSecretWithNotifications};
